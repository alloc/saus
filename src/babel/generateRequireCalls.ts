import { t } from './index'

export function generateRequireCalls(
  node: t.ImportDeclaration,
  requireIdent: string,
  requireCalls: string[],
  moduleId = node.source.value,
  moduleType?: 'esm' | 'cjs'
) {
  const requireCall = `await ${requireIdent}("${moduleId}")`
  const bindings: string[] = []

  let needsImportStar = false
  let needsImportDefault = false

  for (const spec of node.specifiers) {
    const alias = spec.local.name
    if (t.isImportNamespaceSpecifier(spec)) {
      needsImportStar = true
      requireCalls.push(
        moduleType == 'cjs'
          ? `const ${alias} = ${requireCall};\n`
          : moduleType == 'esm'
          ? `const { ...${alias} } = ${requireCall}; ` +
            `delete ${alias}.default;\n`
          : `const ${alias} = __importStar(${requireCall});\n`
      )
    } else if (t.isImportDefaultSpecifier(spec)) {
      if (moduleType !== 'esm') {
        needsImportDefault = true
        requireCalls.push(`const ${alias} = __importDefault(${requireCall});\n`)
      } else {
        bindings.push(`default: ${alias}`)
      }
    } else {
      bindings.push(
        spec.imported.start == spec.local.start
          ? alias
          : t.isIdentifier(spec.imported)
          ? `${spec.imported.name}: ${alias}`
          : `["${spec.imported.value.replace(/"/g, '\\"')}"]: ` + alias
      )
    }
  }

  if (bindings.length) {
    const list = bindings.join(', ')
    requireCalls.push(`const { ${list} } = ${requireCall};\n`)
  } else if (!node.specifiers.length) {
    requireCalls.push(`${requireCall};\n`)
  }

  return { needsImportStar, needsImportDefault }
}
