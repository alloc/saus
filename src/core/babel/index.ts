export * from './babel'
export * from './config'
export * from './exports'
export * from './imports'
export * from './program'
export * from './queries'
export * from './resolveReferences'
export * from './transform'
