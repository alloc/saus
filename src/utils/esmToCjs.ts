import { MagicString, NodePath, t } from '../babel'
import createDebug from 'debug'

const debug = createDebug('saus:commonjs')

const exportKeyword = 'export'

/**
 * Rewrite exports to CommonJS.
 */
export function esmExportsToCjs(
  program: NodePath<t.Program>,
  editor: MagicString,
  requireId = 'require'
) {
  for (const path of program.get('body')) {
    if (!path.isExportDeclaration()) {
      continue
    }
    const { start, end } = path.node as {
      start: number
      end: number
    }
    const decl = path.get('declaration') as NodePath<t.Declaration>
    if (path.isExportNamedDeclaration()) {
      const { node } = path
      if (node.source) {
        const requireCall = `${requireId}("${node.source.value}")`
        if (t.isExportNamespaceSpecifier(node.specifiers[0])) {
          const { name } = node.specifiers[0].exported
          editor.overwrite(start, end, `exports.${name} = ${requireCall}`)
        } else {
          const required = program.scope.generateUid()
          editor.overwrite(start, end, `const ${required} = ${requireCall}`)
          for (const spec of node.specifiers as t.ExportSpecifier[]) {
            const property = t.isStringLiteral(spec.exported)
              ? `["${spec.exported.value.replace(/"/g, '\\"')}"]`
              : `.${spec.exported.name}`

            editor.appendLeft(
              end,
              `\nexports.${spec.local.name} = ${required}${property}`
            )
          }
        }
      } else if (node.specifiers.length) {
        editor.overwrite(
          start,
          end,
          node.specifiers
            .map(spec =>
              t.isExportSpecifier(spec)
                ? (t.isStringLiteral(spec.exported)
                    ? `exports["${spec.exported.value.replace(/"/g, '\\"')}"]`
                    : `exports.${spec.exported.name}`) + ` = ${spec.local.name}`
                : false
            )
            .filter(Boolean)
            .join('\n')
        )
      } else if (decl.isFunctionDeclaration() || decl.isClassDeclaration()) {
        const { name } = decl.node.id!
        editor.overwrite(
          start,
          start + exportKeyword.length,
          `exports.${name} =`
        )
      } else if (decl.isVariableDeclaration()) {
        const { kind, declarations } = decl.node
        if (declarations.length > 1) {
          throw Error(
            `Multiple declarators in "export ${kind}" statement is unsupported`
          )
        }
        editor.overwrite(start, start + `export ${kind} `.length, `exports.`)
      }
    } else if (path.isExportAllDeclaration()) {
      const { node } = path
      const requireCall = `${requireId}("${node.source.value}")`
      editor.overwrite(
        start,
        end,
        // The `default` export is not set by `export *` statements.
        `Object.assign(exports, ${requireCall}, {default: exports.default})`
      )
    } else if (path.isExportDefaultDeclaration()) {
      editor.overwrite(
        start,
        start + `export default `.length,
        `exports.default = `
      )
    }
  }
}
