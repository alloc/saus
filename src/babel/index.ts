import * as babel from '@babel/core'
import { types as t, NodePath } from '@babel/core'
import MagicString, { Bundle as MagicBundle } from 'magic-string'

export { babel, t, NodePath, MagicString, MagicBundle }

const toArray = <T>(arg: T): T extends any[] ? T : T[] =>
  Array.isArray(arg) ? arg : ([arg] as any)

export function resolveReferences(
  rootPaths: NodePath | NodePath[],
  filter = (_path: NodePath) => false
): NodePath<t.Statement>[] {
  const crawled = new Set<NodePath>()
  const referenced = new Set<NodePath<t.Statement>>()

  toArray(rootPaths).forEach(crawl)

  function crawl(basePath: NodePath) {
    crawled.add(basePath)
    basePath.traverse({
      JSXIdentifier: onIdentifier,
      Identifier: onIdentifier,
    })
  }

  function onIdentifier(path: NodePath<t.Identifier | t.JSXIdentifier>) {
    const { parentPath } = path
    if (parentPath.isJSXClosingElement() || isPropertyName(path)) {
      return
    }

    const { name } = path.node
    if (path.isJSXIdentifier() && /^[a-z]/.test(name)) {
      return
    }

    const binding = path.scope.getBinding(name)
    if (!binding) {
      return // Global or undeclared variable
    }

    const bindPath = binding.path
    if (getFirstAncestor(bindPath, p => referenced.has(p as any))) {
      return // Inside a referenced statement
    }

    addReference(bindPath)

    if (
      crawled.has(bindPath) ||
      getFirstAncestor(bindPath, p => crawled.has(p))
    ) {
      return // Already crawled
    }

    crawl(bindPath)
  }

  function addReference(path: babel.NodePath | null | undefined) {
    if (path && filter(path)) {
      path = path.getStatementParent()
      if (path) {
        referenced.add(path as NodePath<t.Statement>)
      }
    }
  }

  return Array.from(referenced).sort((a, b) => {
    return a.node.start! - b.node.start!
  })
}

/**
 * Is the given node path eitherâ€¦
 *   - the name of an object property being declared
 *   - the name of a property being accessed
 */
export function isPropertyName({ parentKey, parentPath }: NodePath) {
  return (
    parentPath &&
    ((parentPath.isObjectProperty() &&
      !parentPath.node.computed &&
      parentKey === 'key') ||
      (parentPath.isMemberExpression() &&
        !parentPath.node.computed &&
        parentKey === 'property'))
  )
}

export function removeSSR(): babel.Visitor {
  return {
    MetaProperty(path) {
      if (!path.get('meta').isIdentifier({ name: 'import' })) return
      const expr = path.findParent(p => !p.parentPath?.isMemberExpression())!
      if (expr.toString() == 'import.meta.env.SSR') {
        const parent = expr.parentPath!
        if (expr.parentKey == 'test') {
          if (parent.isIfStatement() || parent.isConditionalExpression()) {
            if (parent.node.alternate) {
              parent.replaceWith(parent.node.alternate)
            } else {
              parent.remove()
            }
          } else {
            expr.replaceWith(t.booleanLiteral(false))
          }
        } else if (
          expr.parentKey == 'left' &&
          parent.equals('operator', '&&')
        ) {
          parent.replaceWith(t.booleanLiteral(false))
        } else {
          expr.replaceWith(t.booleanLiteral(false))
        }
      }
    },
  }
}

export function isChainedCall(
  path: NodePath
): path is NodePath<t.MemberExpression> {
  return (
    path.isMemberExpression() &&
    path.parentKey === 'callee' &&
    path.get('object').isCallExpression()
  )
}

export function flattenCallChain(
  path: NodePath,
  calls: NodePath<t.CallExpression>[] = []
) {
  if (path.isCallExpression()) {
    flattenCallChain(path.get('callee'), calls)
    calls.push(path)
  } else if (isChainedCall(path)) {
    flattenCallChain(path.get('object'), calls)
    calls.push(path.parentPath as NodePath<t.CallExpression>)
  }
  return calls
}

export function getImportDeclaration(
  program: NodePath<t.Program>,
  moduleName: string
) {
  return program
    .get('body')
    .find(
      stmt => stmt.isImportDeclaration() && stmt.node.source.value == moduleName
    ) as NodePath<t.ImportDeclaration> | undefined
}

export function getImportDeclarations(program: NodePath<t.Program>) {
  return program
    .get('body')
    .filter(stmt =>
      stmt.isImportDeclaration()
    ) as NodePath<t.ImportDeclaration>[]
}

export function getFirstAncestor(
  path: NodePath,
  test: (path: NodePath) => boolean
) {
  let parentPath = path.parentPath
  while (parentPath && !test(parentPath)) {
    parentPath = parentPath.parentPath
  }
  return parentPath
}

export function inferSyntaxPlugins(filename: string): babel.PluginItem[] {
  return /\.tsx?$/.test(filename)
    ? [['@babel/syntax-typescript', { isTSX: filename.endsWith('x') }]]
    : []
}

export function transformSync(
  code: string,
  filename: string,
  config: babel.TransformOptions | babel.PluginItem[]
) {
  if (Array.isArray(config)) {
    config = { plugins: config }
  }
  const syntaxPlugins = inferSyntaxPlugins(filename)
  if (syntaxPlugins.length) {
    config.plugins = syntaxPlugins.concat(config.plugins || [])
  }
  return babel.transformSync(code, {
    filename,
    babelrc: false,
    configFile: false,
    sourceMaps: true,
    ...config,
  })
}

/** Remove a `NodePath`, its preceding whitespace, and its trailing newline (if one exists). */
export function remove(path: NodePath, source: MagicString) {
  let start = path.node.start!
  let end = path.node.end!
  start = getWhitespaceStart(start, source.original)
  end = getTrailingLineBreak(end, source.original)
  source.remove(start, end)
}

export function getWhitespaceStart(start: number, source: string) {
  while (source[start - 1] === ' ') start--
  return start
}

export function getTrailingLineBreak(end: number, source: string) {
  return source[end] === '\n' ? end + 1 : end
}

export function isConsoleCall(path: NodePath) {
  if (!path.isCallExpression()) return false
  const callee = path.get('callee')
  return (
    callee.isMemberExpression() &&
    callee.get('object').isIdentifier({ name: 'console' })
  )
}
