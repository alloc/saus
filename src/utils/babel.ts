import * as babel from '@babel/core'
import { NodePath, types as t } from '@babel/core'

export { getBabelConfig } from './babel/config'
export { getBabelProgram } from './babel/program'
export { transformAsync, transformSync } from './babel/transform'
export { NodePath, t, babel }
