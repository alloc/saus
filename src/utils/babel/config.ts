import * as babel from '@babel/core'

export function getBabelConfig(
  filename: string,
  config: babel.TransformOptions | babel.PluginItem[] = {}
): babel.TransformOptions {
  if (Array.isArray(config)) {
    config = { plugins: config }
  }
  const syntaxPlugins = inferSyntaxPlugins(filename)
  if (syntaxPlugins.length) {
    config.plugins = syntaxPlugins.concat(config.plugins || [])
  }
  return {
    filename,
    babelrc: false,
    configFile: false,
    sourceMaps: true,
    ...config,
  }
}

let babelTypeScriptSyntax: string

export function inferSyntaxPlugins(filename: string): babel.PluginItem[] {
  if (/\.tsx?$/.test(filename)) {
    babelTypeScriptSyntax ||= require.resolve('@babel/plugin-syntax-typescript')
    return [[babelTypeScriptSyntax, { isTSX: filename.endsWith('x') }]]
  }
  return []
}
