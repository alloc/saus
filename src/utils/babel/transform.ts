import * as babel from '@babel/core'
import { NodePath, transformFromAstSync, types as t } from '@babel/core'
import MagicString from 'magic-string'
import { getBabelConfig } from './config'
import { getTrailingLineBreak, getWhitespaceStart } from './queries'

export const transformSync = (
  code: string,
  filename: string,
  config: babel.TransformOptions | babel.PluginItem[]
) => babel.transformSync(code, getBabelConfig(filename, config))

export const transformAsync = (
  code: string,
  filename: string,
  config: babel.TransformOptions | babel.PluginItem[]
) => babel.transformAsync(code, getBabelConfig(filename, config))

export function removeSSR(source: MagicString): babel.Visitor {
  return {
    MetaProperty(path) {
      if (!path.get('meta').isIdentifier({ name: 'import' })) return
      const expr = path.findParent(p => !p.parentPath?.isMemberExpression())!
      if (expr.toString() == 'import.meta.env.SSR') {
        const parent = expr.parentPath!
        if (expr.parentKey == 'test') {
          if (parent.isIfStatement() || parent.isConditionalExpression()) {
            if (parent.node.alternate) {
              replaceWith(parent, parent.node.alternate, source)
            } else {
              remove(parent, source)
            }
          } else {
            replaceWith(expr, 'false', source)
          }
        } else if (
          expr.parentKey == 'left' &&
          parent.equals('operator', '&&')
        ) {
          replaceWith(parent, 'false', source)
        } else {
          replaceWith(expr, 'false', source)
        }
      }
    },
  }
}

export function replaceWith(
  path: NodePath,
  replacement: babel.Node | string,
  source: MagicString
) {
  const [start, end] = getExpandedRange(path, source)
  if (typeof replacement === 'string') {
    path.replaceWithSourceString(replacement)
    source.overwrite(start, end, replacement)
  } else {
    let stmts: t.Statement[]
    if (t.isBlockStatement(replacement)) {
      stmts = replacement.body
    } else if (t.isStatement(replacement)) {
      stmts = [replacement]
    } else if (t.isExpression(replacement)) {
      stmts = [t.expressionStatement(replacement)]
    } else {
      throw new Error(`Node type not supported by "replaceWith"`)
    }
    const { code: replacementString } = transformFromAstSync(t.program(stmts))!
    source.overwrite(start, end, replacementString!)
    path.replaceWith(replacement)
  }
}

/** Remove a `NodePath`, its preceding whitespace, and its trailing newline (if one exists). */
export function remove(path: NodePath, source: MagicString) {
  const [start, end] = getExpandedRange(path, source)
  source.remove(start, end)
  path.remove()
}

function getExpandedRange(path: NodePath, source: MagicString) {
  let start = path.node.start!
  let end = path.node.end!
  if (path.node.leadingComments) {
    start = path.node.leadingComments.reduce(
      (start, comment) => Math.min(start, comment.start!),
      start
    )
  }
  start = getWhitespaceStart(start, source.original)
  end = getTrailingLineBreak(end, source.original)
  return [start, end] as const
}
