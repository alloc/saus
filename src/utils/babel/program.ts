import { NodePath, types as t } from '@babel/core'
import { transformSync } from './transform'

export function getBabelProgram(source: string, filename: string) {
  let program: NodePath<t.Program> | undefined
  const visitor: babel.Visitor = {
    Program: path => {
      program = path
      path.stop()
    },
  }

  // Use the AST for traversal, but not code generation.
  transformSync(source, filename, {
    plugins: [{ visitor }],
    sourceMaps: false,
    sourceType: 'module',
    code: false,
  })

  return program!
}
