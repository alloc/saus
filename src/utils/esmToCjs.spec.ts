import endent from 'endent'
import { describe, expect, it } from 'vitest'
import { MagicString, getBabelProgram } from '../babel'
import { esmExportsToCjs } from './esmToCjs'

describe('esmExportsToCjs', () => {
  describe('named exports', () => {
    it('rewrites functions', () => {
      expect(transform(`export function foo() {}`)).toMatchSnapshot()
    })
    it('rewrites classes', () => {
      expect(transform(`export class Foo {}`)).toMatchSnapshot()
    })
    it('rewrites bindings', () => {
      expect(transform(`export const foo = 1`)).toMatchSnapshot()
      expect(transform(`export let foo = 2`)).toMatchSnapshot()
    })
    it('rewrites identifiers', () => {
      const code = endent`
        const Foo = 1, Bar = 2
        export { Foo, Bar as default }
      `
      expect(transform(code)).toMatchSnapshot()
    })
  })
})

function transform(code: string) {
  const program = getBabelProgram(code, 'test.js')
  const editor = new MagicString(code)
  esmExportsToCjs(program, editor)
  return editor.toString()
}
