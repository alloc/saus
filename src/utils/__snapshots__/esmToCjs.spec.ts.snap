// Vitest Snapshot v1

exports[`esmToCjs > named exports > rewrites bindings 1`] = `"exports.foo = 1"`;

exports[`esmToCjs > named exports > rewrites bindings 2`] = `"exports.foo = 2"`;

exports[`esmToCjs > named exports > rewrites classes 1`] = `"exports.Foo = class Foo {}"`;

exports[`esmToCjs > named exports > rewrites functions 1`] = `"exports.foo = function foo() {}"`;

exports[`esmToCjs > named exports > rewrites identifiers 1`] = `
"const Foo = 1, Bar = 2
exports.Foo = Foo
exports.default = Bar"
`;
